lib/segment/src/index/query_optimization/rescore_formula/evaluate.rs

```diff
@@ -0,0 +1,224 @@
+use std::collections::HashMap;
+use std::ops::Neg;
+
+use common::types::ScoreType;
+use geo::{Distance, Haversine};
+use serde_json::Value;
+
+use super::parsed_formula::{Expression, Operation, VariableId};
+use crate::common::operation_error::{OperationError, OperationResult};
+use crate::json_path::JsonPath;
+use crate::types::GeoPoint;
+
+const DEFAULT_SCORE: ScoreType = 0.0;
+
+/// Ephemeral type to hold all formula variables for a single point
+pub(super) struct PointVariables<'a> {
```

Comment:
I think this structure is a bit redundent, we can evaluate expression without explicilty saving all intemediate values into this structure, instead we can compute final value in `FormulaScorer`

Reply:
This makes a lot of sense to skip the work when multiplying by zero, for example. 

Now implemented in https://github.com/qdrant/qdrant/pull/5980/commits/6d5edbc795ca56d06669c878602a49ee07ba9ac3